### PR TITLE
[feat] add MXFP8 support on H3

### PR DESCRIPTION
**Pull request by Davids048**
## Purpose

Add Blackwell MXFP8 inference for the two linear layers in every MiniMax-H3 transformer feed-forward network
(FFN). The FFN accepts BF16 hidden states, executes both matrix multiplications with MXFP8 values and E8M0
block scales, and returns BF16 hidden states.

## Changes

- Add Blackwell MXFP8 inference for MiniMax-H3 FFN `fc_in` and `fc_out` layers. Weights are quantized after
  checkpoint loading and LoRA merging.
- Add support for MXFP8 quantization and configs.
- Simplify FastH3 inference entry points and add Nsight Systems profilers and NVTX ranges for MiniMax-H3.

The FFN dtype flow is:

```text
BF16 hidden states
  -> fused MXFP8 quantization + scale swizzle
  -> MXFP8 fc_in GEMM
  -> BF16 packed SwiGLU input
  -> fused SwiGLU + MXFP8 quantization + scale swizzle
  -> MXFP8 fc_out GEMM
  -> BF16 FFN output
```

Enable the path with:

```python
QuantizationConfig(transformer_quant="MXFP8")
```

## Test Plan

Profile one 345-frame FastH3 generation after three warmup generations:

```bash
bash examples/inference/basic/basic_fasth3_simplified_profile.sh
```

The MXFP8 run used this command. The BF16 totals came from predecessor run `20260831T004438Z-4009913`, which
used the same profile settings with transformer quantization disabled.

Profile configuration:

- 4 NVIDIA GB200 GPUs with sequence parallelism 4
- 1344x768, 345 frames, 24 frames per second
- FastH3 `vsa-datafree` LoRA, 5 scheduler points, seed 1000
- Video sparse attention at 90% sparsity with 64-token query/key tiles
- Video variational autoencoder compilation enabled; diffusion transformer compilation disabled
- 800 FFN NVTX ranges across four GPUs

## Test Results

Summed GPU kernel time inside the same 800 FFN NVTX ranges:

| FFN stage                    | BF16 total   | MXFP8 total  | Speedup |
| ---------------------------- | -----------: | -----------: | ------: |
| Input quantization           |            - |   260.859 ms |       - |
| `fc_in`                      | 3,491.160 ms | 1,902.536 ms |  1.835x |
| SwiGLU + output quantization |   373.079 ms |   786.248 ms |  0.475x |
| `fc_out`                     | 1,852.935 ms |   983.061 ms |  1.885x |
| **All FFN kernels**          | 5,717.174 ms | 3,932.703 ms |  1.454x |

MXFP8 reduces summed FFN kernel time by 31.21% in this profile. For the single measured generation in each
format, output-tensor production took 17.84 seconds with MXFP8 and 18.15 seconds with BF16. Including frame
postprocessing and media saving, elapsed latency was 19.33 seconds with MXFP8 and 19.53 seconds with BF16. One
measurement per format does not establish run-to-run variability. The run artifacts do not record a commit ID
or source hashes, so the exact profiled source snapshot is unverified.

## Checklist

- [ ] I ran `pre-commit run --all-files` and fixed all issues
- [ ] I added or updated tests for my changes
- [ ] I updated documentation if needed
- [ ] I considered GPU memory impact of my changes

**For model/pipeline changes:**

- [ ] I verified SSIM regression tests pass
- [x] The support matrix does not need an update because this PR adds a quantization path for MiniMax-H3.

### CODE REVIEW

**Comment by SolitaryThinker** [comment]
Rebased onto current origin/main to pick up the main-side FastH3 loader/CLI and unit-test fixes; no content changes (range-diff clean, with only the already-upstreamed h3-sequential-load hunk absorbed by main).

File: fastvideo/models/loader/fsdp_load.py [hunk 1]
```diff
@@ -287,7 +301,7 @@ def maybe_load_fsdp_model(
     # responsibility is just to materialize the quantized weight buffers
     # from the freshly-loaded bf16 weights. No-op when no quantized layers
     # are present (lazy imports inside the helper).
-    _maybe_quantize_model(model)
+    _maybe_quantize_model(model, defer_weight_conversion_until_lora_merge=lora_path is not None)
```
Comment (SolitaryThinker) [inline fastvideo/models/loader/fsdp_load.py:326]: **P1:** This quantizes after `shard_model` has converted the H3 FFN weights into sharded `DTensor`s. NVFP4 then packs only `weight.to_local()` into ordinary non-persistent buffers, while MXFP8 passes the sharded weight to Quack. These packed buffers are not managed or all-gathered by FSDP, so quantized FSDP inference can use rank-local/incomplete weights. Please reject NVFP4/MXFP8 with FSDP inference before sharding, or implement packing and gathering with a correct distributed-buffer lifecycle.

File: fastvideo/pipelines/lora_pipeline.py [hunk 2]
```diff
@@ -532,6 +550,7 @@ def set_lora_adapter(self,
                                lora_path)
             if unmatched:
                 logger.warning("LoRA adapter %s: %d weights did not reach a layer", lora_path, len(unmatched))
+        _convert_quantized_weights_after_lora_merge(self.trainable_transformer_modules)
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/lora_pipeline.py:590]: **P1:** This is a one-time repack after the initial merge, but the public unmerge and adapter-switching paths remain enabled. Default NVFP4 conversion purges `base_layer.weight`, so `unmerge_lora_weights()` later dereferences `weight.data` and crashes. If the original is retained (or with MXFP8), the packed buffer still contains the merged adapter after unmerge, producing stale/wrong behavior. Please either reject these lifecycle operations explicitly for packed models, or restore/repack consistently on every merge/unmerge and cover adapter swap/unmerge in tests.

File: fastvideo/layers/quantization/mxfp8_config.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        set_weight_attrs(weight, {"input_dim": 1, "output_dim": 0})
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(weight, extra_weight_attrs)
+        layer.register_buffer("_mxfp8_weight", None, persistent=False)
+        layer.register_buffer("_mxfp8_weight_scale", None, persistent=False)
+
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        """Prequantize one adapter-merged BF16 linear weight."""
+        from fastvideo.layers.mxfp8linear import quantize_mxfp8_weight_blockwise
+
```
Comment (SolitaryThinker) [inline fastvideo/layers/quantization/mxfp8_config.py:53]: **P1:** This PR does not satisfy the add-model component gate because non-skip numerical parity for the new MXFP8 FFN is missing. The risk is silent numerical drift in the custom Triton activation/SwiGLU quantization and scaled GEMM path. Please add Blackwell-gated tests comparing weight/activation quantization and FFN outputs against BF16/Quack, plus H3 generated-output or SSIM quality evidence.

File: fastvideo/layers/mxfp8linear.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
+# SPDX-License-Identifier: Apache-2.0
+"""MXFP8 block quantization and linear operations for Blackwell inference."""
+
+from __future__ import annotations
+
+import torch
+import torch.nn.functional as F
+import triton
+import triton.language as tl
+from quack.mx_utils import to_blocked, to_mx
```
Comment (SolitaryThinker) [inline fastvideo/layers/mxfp8linear.py:10]: **P2:** `quack.mx_utils` is a runtime dependency for MXFP8, but `quack-kernels` is not declared directly in the project dependencies or feature extra. A standard install can select MXFP8 and fail at import, and relying on a transitive FlashAttention dependency is brittle. Please add a pinned direct feature dependency and preferably keep the missing-backend error lazy and actionable.

**Review by SolitaryThinker — CHANGES_REQUESTED** [review]
Requesting changes based on the implementation review; CI-only failures are intentionally omitted. The main blockers are distributed/FSDP correctness, LoRA lifecycle correctness, and missing non-skip MXFP8 parity. I also left a dependency-declaration comment. I reviewed the component, loader, LoRA integration, and quality/evidence lanes. Blackwell numerical validation could not be run in my local environment.

File: fastvideo/layers/quantization/mxfp8_config.py (hunk 3 above)
Comment (Davids048) [reply fastvideo/layers/quantization/mxfp8_config.py:53]: I added Blackwell-gated numerical parity tests in fastvideo/tests/ops/quantization/test_mxfp8.py.

| Coverage                              | Comparison                                       | Result                                               |
| ------------------------------------- | ------------------------------------------------ | ---------------------------------------------------- |
| Weight quantization                   | FastVideo versus Quack                           | Exact FP8 values and E8M0 scale bytes                |
| Activation quantization and swizzling | Fused Triton versus Quack                        | Exact FP8 values and E8M0 scale bytes                |
| Fused SwiGLU quantization             | Fused Triton versus BF16 SwiGLU followed by Quack | Exact FP8 values and E8M0 scale bytes                |
| Complete H3 FFN                       | MXFP8 versus BF16                                 | Cosine similarity 0.99727; relative L2 error 0.07538 |

  The complete FFN test requires cosine similarity > 0.995 and relative L2 error < 0.10. All four tests execute on
  Blackwell and pass on GB200:

  4 passed

  I also generated 20 matched videos using the same prompts, seeds, resolutions, frame counts, and inference
  settings under BF16 and MXFP8. Side-by-side review showed no visible quality difference between the BF16 and MXFP8 outputs.

MXFP8


https://github.com/user-attachments/assets/d78ad1f3-46cb-4f62-8325-9ce02490ae3e



BF16


https://github.com/user-attachments/assets/664e70e7-2bcc-4370-a0cd-8f7a47f66c65

File: fastvideo/pipelines/lora_pipeline.py (hunk 2 above)
Comment (Davids048) [reply fastvideo/pipelines/lora_pipeline.py:590]: rejected them explicitly, added issue #1821

File: fastvideo/models/loader/fsdp_load.py (hunk 1 above)
Comment (Davids048) [reply fastvideo/models/loader/fsdp_load.py:326]: rejected the quantization modes with FSDP inference currently. Added issue to support them in a later PR. Issue #1822

File: fastvideo/layers/quantization/mxfp8_config.py (hunk 3 above)
Comment (SolitaryThinker) [reply fastvideo/layers/quantization/mxfp8_config.py:53]: The Blackwell-gated exact quantization checks and full-FFN BF16 comparison close the numerical parity concern, and the reported 20 matched generations cover the requested output-quality check. The MXFP8 attachment currently returns 404 while the BF16 attachment loads; please reattach the MXFP8 sample for the durable record. I am treating that link repair as non-blocking given the recorded GB200 results.

**Review by SolitaryThinker — APPROVED** [review]
Re-reviewed the latest head against both origin/main and the prior reviewed commit. All four requested changes are addressed: the FastH3 extra directly pins Quack, unsupported FSDP-inference quantization is rejected before model construction, the packed-weight LoRA lifecycle is explicitly guarded with MXFP8 adapter changes requantized, and Blackwell numerical parity now covers activation/scales, fused SwiGLU, and the full H3 FFN. The author also reported a 4-pass GB200 run and 20 matched BF16/MXFP8 generations with no visible quality regression. No new blocking code findings. Per maintainer direction, CI status was not considered. I could not independently rerun Blackwell kernels locally; the MXFP8 sample attachment currently returns 404 and should be reattached for the record, but this is non-blocking given the recorded numerical and visual evidence.
